---
sidebar_position: 1
---

# Introduction

**Thank you for using SEGUL!** üôèüèª

We develop **SEGUL (SEquence and Genomic UtiLities)** to address the need for a high-performance and accessible phylogenomic tool. It is particularly well-suited for large-scale phylogenomic projects, especially those involving thousands of loci and hundreds of samples. Additionally, SEGUL is capable of handling small Sanger sequences effectively.

:::info
[PySEGUL](https://pypi.org/project/pysegul/), a Python binding of SEGUL API, is available in beta on PyPI. You can install it using pip:

```bash
pip install pysegul
```

Learn more about using SEGUL API in Python [here](/docs/api-usage/python/intro).
:::

## Citation

> Handika, H., and J. A. Esselstyn. 2024. SEGUL: Ultrafast, memory-efficient and mobile-friendly software for manipulating and summarizing phylogenomic datasets. _Molecular Ecology Resources_. [https://doi.org/10.1111/1755-0998.13964](https://doi.org/10.1111/1755-0998.13964).

If you are interested in reading the journal article but do not have access to it, you can request a copy from the authors on [ResearchGate](https://www.researchgate.net/publication/380125709_SEGUL_Ultrafast_memory-efficient_and_mobile-friendly_software_for_manipulating_and_summarizing_phylogenomic_datasets).

## Navigating the Documentation

We provide quick start [below](#quick-start-%EF%B8%8F) to help you get started with SEGUL. The quick start assumes familiarity with phylogenomic data and some experience using related tools. A detailed instruction on installing SEGUL is available in the [installation guidelines](/docs/installation/overview). If you are using the GUI app, we recommend checking the [GUI general usage](/docs/gui-usage/general) section. For using the CLI application, we recommend starting with the [CLI Overview](/docs/cli-usage/overview) section. These sections provide an overview of the app, input and output formats, datatype, and other general information about the app. If you have any questions, issues, and feedback about the app and its documentation, please refer to the [support](/docs/support) section. We also welcome for collaboration and contribution.

:::tip
If you are reading the documentation on mobile devices, at the top of each page, you can find a dropdown menu that allows you to navigate to different sections of the documentation. For larger screen reading, the sidebar on the right side of the page provides the same functionality.
:::

## Quick Start ‚è±Ô∏è

### Platform Support

#### Desktop

| Platform                          | GUI | CLI |
| --------------------------------- | --- | --- |
| Linux                             | ‚úÖ  | ‚úÖ  |
| MacOS                             | ‚úÖ  | ‚úÖ  |
| Windows                           | ‚úÖ  | ‚úÖ  |
| Windows Subsystem for Linux (WSL) | ‚ùå  | ‚úÖ  |

#### Mobile

| Platform | GUI | CLI |
| -------- | --- | --- |
| iOS      | ‚úÖ  | ‚ùå  |
| iPadOS   | ‚úÖ  | ‚ùå  |
| Android  | ‚úÖ  | ‚ùå  |

### CLI vs GUI vs API

For general users, the choice will be between the command-line (CLI) and graphical interface (GUI) applications. If your answer is yes to any of the following questions, you should use the CLI version of the app. Otherwise, either GUI or CLI will serve your need. You can also install both versions and use them interchangeably. They each operates in their own contained environment and do not interfere with the other.

1. Are you planning to run the app on non-GUI environment, such as a [HPC](https://www.ibm.com/topics/hpc) cluster?
2. Do you run SEGUL as part of a pipeline?
3. Do you you need the utmost efficiency?

Learn more about the differences between the CLI and GUI version [here](./cli_gui.md).

For developer, SEGUL is available as a Rust crate, which can be easily integrated in Rust code. The Python library is available as a Python binding to SEGUL API called PySEGUL. It is user-friendly and does not require Rust knowledge to use it. The Rust crate also works with R. However, using it in R requires some Rust knowledge.

Learn more about using SEGUL API [here](/docs/category/api-usages).

:::tip
If you write a pipeline in Python, we recommend using PySEGUL. It is available in Python Package Index (PyPI) that will simplify dependency management.
:::

### Installation

For GUI apps, we recommend downloading from the [official store](/docs/installation/install_gui) of your operating system. The fastest installation route for CLI is using the pre-compiled binaries. If you are familiar with installing single executable CLI apps, you can download the latest release using the links below. The [installation guide](/docs/installation/install_binary) provides more detailed instructions.

| Platform              | Link                                                                                                  | Description                                                                                                               |
| --------------------- | ----------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------- |
| Linux ARM64           | [download](https://github.com/hhandika/segul/releases/latest/download/segul-Linux-arm64.tar.gz)       | For Linux on ARM64 architecture (uncommon).                                                                               |
| Linux x86_64 (static) | [download](https://github.com/hhandika/segul/releases/latest/download/segul-Linux-musl-x86_64.tar.gz) | For Linux with old [GLIBC](https://www.gnu.org/software/libc/) version. Common in HPC clusters.                           |
| Linux x86_64          | [download](https://github.com/hhandika/segul/releases/latest/download/segul-Linux-x86_64.tar.gz)      | For Linux with modern [GLIBC](https://www.gnu.org/software/libc/) version. Most recent Linux distribution, including WSL. |
| MacOS ARM64           | [download](https://github.com/hhandika/segul/releases/latest/download/segul-macOS-x86_64.tar.gz)      | For MacOS on Apple M series.                                                                                              |
| MacOS x86_64          | [download](https://github.com/hhandika/segul/releases/latest/download/segul-macOS-x86_64.tar.gz)      | For MacOS on Intel.                                                                                                       |
| Windows x86_64        | [download](https://github.com/hhandika/segul/releases/latest/download/segul-Windows-x86_64.zip)       | Most Windows devices. May work on ARM Windows as well.                                                                    |

### Usage Overview

SEGUL has a growing list of features to help you manipulate and summarize your phylogenomic datasets.

#### Feature Quick Links

| Feature                        | Quick Link                                                           |
| ------------------------------ | -------------------------------------------------------------------- |
| Alignment concatenation        | [CLI](./cli-usage/concat) / [GUI](./gui-usage/align-concat)          |
| Alignment conversion           | [CLI](./cli-usage/convert) / [GUI](./gui-usage/align-convert)        |
| Alignment filtering            | [CLI](./cli-usage/filter) / [GUI](./gui-usage/align-filter)          |
| Alignment splitting            | [CLI](./cli-usage/split) / [GUI](./gui-usage/align-split)            |
| Alignment partition conversion | [CLI](./cli-usage/partition) / [GUI](./gui-usage/align-partition)    |
| Alignment summary statistics   | [CLI](./cli-usage/summary) / [GUI](./gui-usage/align-summary)        |
| Genomic summary statistics     | [CLI](./cli-usage/genomic) / [GUI](./gui-usage/genomic)              |
| Sequence extraction            | [CLI](./cli-usage/extract) / [GUI](./gui-usage/sequence-extract)     |
| Sequence ID extraction         | [CLI](./cli-usage/id) / [GUI](./gui-usage/sequence-id)               |
| Sequence ID mapping            | [CLI](./cli-usage/map.md) / [GUI](./gui-usage/sequence-id-map)       |
| Sequence ID renaming           | [CLI](./cli-usage/rename) / [GUI](./gui-usage/sequence-rename)       |
| Sequence removal               | [CLI](./cli-usage/remove) / [GUI](./gui-usage/sequence-remove)       |
| Sequence translation           | [CLI](./cli-usage/translate) / [GUI](./gui-usage/sequence-translate) |
| Log file                       | [CLI](./cli-usage/log) / [GUI](./gui-usage/log)                      |

#### Supported File Formats

Supported input formats for **Genomic** tasks:

| File Format | Description                                                              | Supported extensions                    |
| ----------- | ------------------------------------------------------------------------ | --------------------------------------- |
| FASTQ       | For read summary statistics. Support compressed and uncompressed format. | `.fastq`, `.fq`                         |
| FASTA       | For contig summary statistics.                                           | `.fasta`, `.fa`, `.fna`, `.fsa`, `.fas` |

Supported input and output file formats for **Alignment** and **Sequence** tasks:

| File Format | Description                                                                                                                                                       | Supported extensions                    |
| ----------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------- | --------------------------------------- |
| FASTA       | Include support for interleaved format.                                                                                                                           | `.fasta`, `.fa`, `.fna`, `.fsa`, `.fas` |
| PHYLIP      | Support relaxed-phylip only. Include support for interleaved format. Learn the differences [here](https://biopython.org/docs/1.74/api/Bio.AlignIO.PhylipIO.html). | `.phy`, `.phylip`, `.ph`                |
| NEXUS       | Include support for interleaved format.                                                                                                                           | `.nexus`, `.nex`, `.nxs`                |

Supported input and output partition formats:

| File Format | Description           | Supported extensions          |
| ----------- | --------------------- | ----------------------------- |
| RAxML       | RAxML partition file. | `.txt`, `.part`, `.partition` |
| NEXUS       | NEXUS partition file. | `.nexus`, `.nex`, `.nxs`      |

:::info
SEGUL CLI can handle non-standard file extensions that are not listed above. Use the `--format` option to set the input format. The GUI version will not allow inputting non-standard file extensions. You can change the file extension to one of the supported file extensions or use the CLI version.
:::

#### Example Dataset

We provide sample datasets to help you get started or test the app if you find any issues using your own datasets. The datasets include small alignments in SEGUL supported formats. Due to the large file sizes, we are unable to provide genomic datasets. You can download the genomic data from public repositories, such as [NCBI SRA](https://www.ncbi.nlm.nih.gov/sra).

| Dataset                    | Link                                                                                |
| -------------------------- | ----------------------------------------------------------------------------------- |
| FASTA                      | [Download](https://github.com/hhandika/segul/raw/main/examples/data-fasta.zip)      |
| NEXUS                      | [Download](https://github.com/hhandika/segul/raw/main/examples/data-nexus.zip)      |
| Relaxed-PHYLIP             | [Download](https://github.com/hhandika/segul/raw/main/examples/data-phylip.zip)     |
| Interleaved relaxed-PHYLIP | [Download](https://github.com/hhandika/segul/raw/main/examples/data-phylip.zip)     |
| Concatenated NEXUS         | [Download](https://github.com/hhandika/segul/raw/main/examples/data-concat-nex.zip) |

### GUI Usage

1. Open the app.
2. Use the navigation bar to select the feature you want to use. For example, if you want to concatenate alignments, click the "Alignments" button.
3. Use the dropdown menu to select the task. For alignment concatenation, select "Concatenate alignments".
4. Add the input files by clicking the "Add file" button. On desktop platforms, you can also input a directory by clicking the "Add directory" button. The app will look for matching files in the directory.
5. All the input files will be displayed in the input tab bar. You can remove the file by clicking the "Remove" button. Removing the file will only remove it from input list and not from the file system.
6. Add the output directory by clicking the "Add output directory" button. On mobile platforms, the directory will be the default directory for the app.
7. For some tasks, you also need to add the parameters. For example, if you want to filter alignments, you need to add the filtering parameters.
8. Click the "Run" button to start the task.
9. Once it's done, the app will display the output in the output tab bar. You can also tab the file to open it in the app file viewer. Current version only support plain text and comma-separated (CSV) data.
10. You can also share the output. There are two share options. The quick share will create a zip file containing the output and share it using the system share sheet. You can also share individual files by clicking the share button on the file output viewer.

:::warning
The mobile version of the GUI has limited capabilities in handling large number of files. Find out more in the [guideline for mobile users](/docs/gui-usage/general#specific-guideline-for-mobile-users).
:::

### CLI Usage

`segul` command is structured as follows:

```bash
segul <command> <subcommand> --option1 value1 --option2 value2
```

For example, to concatenate alignments:

```bash
segul align concat --dir alignments/ --output aln-concat
```

Some arguments are required, while others are optional. For example, the `--output` option is optional. If you do not provide it, the app will use the default output directory. This command will concatenate all the alignments in the `alignments` directory and save the output in the `Align-Concat` directory.

```bash
segul align concat --dir alignments/
```

We recommend using the `segul --help` option to see the available options for each command, the `segul <command> --help` option to see the available options for each subcommand, and the `segul <command> <subcommand> --help` option to see the available options for each subcommand.

See the [command usage](./cli-usage/command_options.md) section for more detailed usage.

:::tip
Since version 0.19.0, you don't need to specify the input format. The app will automatically detect the format based on the file extension. Lower version only allow auto detect format for non-directory input. If you have non-standard file extensions, regardless of the version, use the `--format` option to set the input format.
:::

:::warning
The commands below expect SEGUL version 0.19.0+. If you are using lower version, the subcommand become the command and dir input option required input format.

For example:

```bash
segul alignment concat -d <input-directory>
```

In the version lower than 0.19.0, the command above will be:

```bash
segul concat -d <input-directory> -f <input-format>
```

:::

| Feature                        | Commands                                                           |
| ------------------------------ | ------------------------------------------------------------------ |
| Alignment concatenation        | `segul align concat -d <input-directory>`                          |
| Alignment conversion           | `segul align convert -d <input-directory>`                         |
| Alignment filtering            | `segul align filter -d <input-directory> <filtering-options>`      |
| Alignment splitting            | `segul align split -d <input-directory>`                           |
| Alignment partition conversion | `segul partition convert -d <input-directory>`                     |
| Alignment summary statistics   | `segul align summary -d <input-directory>`                         |
| Contig summary statistics      | `segul contig summary -d <input-directory>`                        |
| Read summary statistics        | `segul read summary -d <input-directory>`                          |
| Sequence extraction            | `segul sequence extract -d <input-directory> <extraction-options>` |
| Sequence ID extraction         | `segul sequence id -d <input-directory>`                           |
| Sequence ID mapping            | `segul sequence id --map -d <input-directory>`                     |
| Sequence ID renaming           | `segul sequence rename -d <input-directory>`                       |
| Sequence removal               | `segul sequence remove -d <input-directory>`                       |
| Sequence translation           | `segul sequence translate -d <input-directory>`                    |
| Main help                      | `segul --help`                                                     |
| Command help                   | `segul <command> --help`                                           |
| Subcommand help                | `segul <command> <subcommand> --help`                              |

### API Usage

SEGUL API is available as a Rust crate. You can use it to develop your own application or integrate it with other programming languages. The API is available on [crates.io](https://crates.io/crates/segul).

To add SEGUL API to your project, you can use the `cargo add` command:

```bash
cd my-project

cargo add segul
```

Or add manually in `Cargo.toml` file:

```toml
[dependencies]
segul = "0.*"
```

If you are interested in using SEGUL API in Python, we provide a Python binding called PySEGUL. You can access SEGUL features like using any python library. No Rust knowledge is needed. Install it using pip:

```bash
pip install pysegul
```

```python
import pysegul

def concat_alignments():
    input_dir = 'tests/data'
    input_format = 'nexus'
    datatype = 'dna'
    output_format = 'fasta'
    partition_format = 'raxml'
    prefix = 'concatenated'
    output_dir = 'tests/output'
    concat = pysegul.AlignmentConcatenation(
        input_format,
        datatype,
        output_dir,
        output_format,
        partition_format,
        prefix
        )
    concat.from_dir(input_dir)
    # For inputting a list of files instead of a directory
    input_paths = ['tests/data/alignment1.nex', 'tests/data/alignment2.nex']
    concat.from_files(input_paths)
```

Most of the PySEGUL features follow the same code pattern, except for features that require specific paramaters, such as alignment filtering, sequence extraction, and sequence removal. For these features, use setter to input files or directory. Then, use the matching method parameters to run the analyses. For example, to extract sequences using regular expression:

```python
import pysegul

def extract_sequences():
    input_dir = 'tests/align-data'
    input_format = 'nexus'
    datatype = 'dna'
    output_format = 'fasta'
    output_dir = 'tests/output'
    extract = pysegul.SequenceExtraction(
        input_format,
        datatype,
        output_dir,
        output_format,
        )
    extract.input_dir = input_dir
    extract.extract_regex("(?i)^(abce)")
```

Learn more about using PySEGUL [here](/docs/api-usage/python/intro).

## Additional Resources

- [API Documentation](https://docs.rs/segul/0.18.1/segul/)
- [GUI Source Code](https://github.com/hhandika/segui)
- [CLI & Rust API Source Code](https://github.com/hhandika/segul)
- [Python API Source Code](https://github.com/hhandika/pysegul)
